# ***Notebook for testing***

In [1]:
import os
import sys
import pandas as pd
sys.path.append("../")
from pyspark.sql import SparkSession

## Correspondence

In [2]:
correspondence = pd.read_csv("../data/tables/abs/postcode_correspondences_2021.csv")
correspondence.head(15)

POSTCODE SA2_CODE_2021     SA2_NAME_2021  RATIO_FROM_TO  \
0      800.0     701011002       Darwin City   1.000000e+00   
1      810.0     701021018           Jingili   5.340360e-02   
2      810.0     701021029           Wanguri   5.363500e-02   
3      810.0     701021024              Moil   5.821800e-02   
4      810.0     701021028           Wagaman   5.894620e-02   
5      810.0     701021010             Alawa   6.095950e-02   
6      810.0     701021027              Tiwi   7.289560e-02   
7      810.0     701021023           Millner   7.473410e-02   
8      810.0     701021016     Coconut Grove   8.395780e-02   
9      810.0     701021026       Rapid Creek   9.522160e-02   
10     810.0     701021013  Brinkin - Nakara   1.000345e-01   
11     810.0     701021025        Nightcliff   1.134344e-01   
12     810.0     701021021        Lyons (NT)   1.745598e-01   
13     812.0     701021021        Lyons (NT)   3.000000e-07   
14     812.0     701021014     Buffalo Creek   1.600000e-06   

   INDIV_TO_REGION_QLTY_INDICATOR OVERALL_QUALITY_INDICATOR  BMOS_NULL_FLAG  
0                            Good                      Poor             0.0  
1                            Poor                      Poor             0.0  
2                            Poor                      Poor             0.0  
3                            Poor                      Poor             0.0  
4                            Poor                      Poor             0.0  
5                            Poor                      Poor             0.0  
6                            Poor                      Poor             0.0  
7                            Poor                      Poor             0.0  
8                            Poor                      Poor             0.0  
9                            Poor                      Poor             0.0  
10                           Poor                      Poor             0.0  
11                           Poor                      Poor             0.0  
12                           Poor                      Poor             0.0  
13                           Poor                      Poor             1.0  
14                           Poor                      Poor             1.0

Postcodes may be matched to multiple SA2 regions, therefore we will choose the region with the highest ratio, i.e. the SA2 region with the highest percentage of population for that postcode. For example, in the table above, you can see the postcode 0810 is matched to 12 different SA2 regions. Of the twelve, we will choose code 701021021 - Lyons (NT) - as about 17% of all people from the 0810 postcode are within this region, which is the highest percentage.

In [3]:
# get indices of max ratio values for each postcode
max_indices = correspondence.groupby("POSTCODE")["RATIO_FROM_TO"].idxmax()

In [4]:
correspondence_filtered = correspondence.loc[max_indices].reset_index(drop=True)
correspondence_filtered.head(5)

POSTCODE SA2_CODE_2021 SA2_NAME_2021  RATIO_FROM_TO  \
0     800.0     701011002   Darwin City       1.000000   
1     810.0     701021021    Lyons (NT)       0.174560   
2     812.0     701021019        Karama       0.257474   
3     820.0     701011008   Stuart Park       0.206533   
4     822.0     702041063   East Arnhem       0.209947   

  INDIV_TO_REGION_QLTY_INDICATOR OVERALL_QUALITY_INDICATOR  BMOS_NULL_FLAG  
0                           Good                      Poor             0.0  
1                           Poor                      Poor             0.0  
2                           Poor                      Poor             0.0  
3                           Poor                      Poor             0.0  
4                           Poor                      Poor             0.0

We can then get rid of all columns except postcode and SA2 code, which will be of use for combining our ABS data with the synthetic data.

In [5]:
correspondence_filtered = correspondence_filtered[["POSTCODE", "SA2_CODE_2021"]]
correspondence_filtered.head(5)

POSTCODE SA2_CODE_2021
0     800.0     701011002
1     810.0     701021021
2     812.0     701021019
3     820.0     701011008
4     822.0     702041063

We can now assign a SA2 code for each customer in our synthetic dataset.

In [6]:
# read in data
# data is seperated by pipe "|", not comma
customers = pd.read_csv("../data/tables/synthetic/tbl_consumer.csv", sep="|")
customers

name                          address state  postcode  \
0        Yolanda Williams       413 Haney Gardens Apt. 742    WA      6935   
1              Mary Smith                  3764 Amber Oval   NSW      2782   
2           Jill Jones MD               40693 Henry Greens    NT       862   
3         Lindsay Jimenez        00653 Davenport Crossroad   NSW      2780   
4       Rebecca Blanchard    9271 Michael Manors Suite 651    WA      6355   
...                   ...                              ...   ...       ...   
499994      Jessica Avila    508 Miranda Overpass Apt. 218   QLD      4400   
499995    Steven Thornton  7913 Schwartz Mission Suite 483   VIC      3097   
499996      Christy Smith   5681 Zachary Mountain Apt. 060   NSW      2756   
499997       Donna Sutton                54140 Jacob Point   VIC      3989   
499998     Hannah Wilkins                61055 Long Valley   NSW      1755   

             gender  consumer_id  
0            Female      1195503  
1            Female       179208  
2            Female      1194530  
3            Female       154128  
4            Female       712975  
...             ...          ...  
499994       Female      1385608  
499995  Undisclosed      1466964  
499996  Undisclosed      1253484  
499997       Female       175005  
499998       Female       746835  

[499999 rows x 6 columns]

In [7]:
customers_merged = pd.merge(customers, correspondence_filtered, left_on="postcode", right_on="POSTCODE", how="left")
customers_merged = customers_merged.drop("POSTCODE", axis=1)
customers_merged.head(5)

name                        address state  postcode  gender  \
0   Yolanda Williams     413 Haney Gardens Apt. 742    WA      6935  Female   
1         Mary Smith                3764 Amber Oval   NSW      2782  Female   
2      Jill Jones MD             40693 Henry Greens    NT       862  Female   
3    Lindsay Jimenez      00653 Davenport Crossroad   NSW      2780  Female   
4  Rebecca Blanchard  9271 Michael Manors Suite 651    WA      6355  Female   

   consumer_id SA2_CODE_2021  
0      1195503           NaN  
1       179208     124011455  
2      1194530     702021055  
3       154128     124011452  
4       712975     509031247

In [8]:
customers_merged["SA2_CODE_2021"].isna().sum()

83181

83,181 missing SA2 codes - maybe impute with mean/median values for state when using ABS data later? Or find similar postcodes or use latitude/longitude data?

In [9]:
# replace NaN values with zeroes (helps with merging later)
customers_merged["SA2_CODE_2021"] = customers_merged["SA2_CODE_2021"].fillna(0)
customers_merged.head(5)

name                        address state  postcode  gender  \
0   Yolanda Williams     413 Haney Gardens Apt. 742    WA      6935  Female   
1         Mary Smith                3764 Amber Oval   NSW      2782  Female   
2      Jill Jones MD             40693 Henry Greens    NT       862  Female   
3    Lindsay Jimenez      00653 Davenport Crossroad   NSW      2780  Female   
4  Rebecca Blanchard  9271 Michael Manors Suite 651    WA      6355  Female   

   consumer_id SA2_CODE_2021  
0      1195503             0  
1       179208     124011455  
2      1194530     702021055  
3       154128     124011452  
4       712975     509031247

## ABS Data Cleaning

In [10]:
abs_df = pd.read_csv("../data/tables/abs/ABS_2021.csv")
abs_df.head(15)

DATAFLOW                       MEDAVG: Median/Average  \
0   ABS:C21_G02_SA2(1.0.0)                     1: Median age of persons   
1   ABS:C21_G02_SA2(1.0.0)   2: Median total personal income ($/weekly)   
2   ABS:C21_G02_SA2(1.0.0)     3: Median total family income ($/weekly)   
3   ABS:C21_G02_SA2(1.0.0)  4: Median total household income ($/weekly)   
4   ABS:C21_G02_SA2(1.0.0)     5: Median mortgage repayment ($/monthly)   
5   ABS:C21_G02_SA2(1.0.0)                    6: Median rent ($/weekly)   
6   ABS:C21_G02_SA2(1.0.0)     7: Average number of persons per bedroom   
7   ABS:C21_G02_SA2(1.0.0)                    8: Average household size   
8   ABS:C21_G02_SA2(1.0.0)                     1: Median age of persons   
9   ABS:C21_G02_SA2(1.0.0)                     1: Median age of persons   
10  ABS:C21_G02_SA2(1.0.0)                     1: Median age of persons   
11  ABS:C21_G02_SA2(1.0.0)                     1: Median age of persons   
12  ABS:C21_G02_SA2(1.0.0)                     1: Median age of persons   
13  ABS:C21_G02_SA2(1.0.0)                     1: Median age of persons   
14  ABS:C21_G02_SA2(1.0.0)                     1: Median age of persons   

                         REGION: Region       REGION_TYPE: Region Type  \
0                        AUS: Australia                 AUS: Australia   
1                        AUS: Australia                 AUS: Australia   
2                        AUS: Australia                 AUS: Australia   
3                        AUS: Australia                 AUS: Australia   
4                        AUS: Australia                 AUS: Australia   
5                        AUS: Australia                 AUS: Australia   
6                        AUS: Australia                 AUS: Australia   
7                        AUS: Australia                 AUS: Australia   
8                 112: Richmond - Tweed  SA4: Statistical Area Level 4   
9              127: Sydney - South West  SA4: Statistical Area Level 4   
10                    10102: Queanbeyan  SA3: Statistical Area Level 3   
11                  10106: Young - Yass  SA3: Statistical Area Level 3   
12                        10304: Orange  SA3: Statistical Area Level 3   
13  11202: Richmond Valley - Hinterland  SA3: Statistical Area Level 3   
14                11501: Baulkham Hills  SA3: Statistical Area Level 3   

          STATE: State  TIME_PERIOD: Time Period  OBS_VALUE  
0       AUS: Australia                      2021       38.0  
1       AUS: Australia                      2021      805.0  
2       AUS: Australia                      2021     2120.0  
3       AUS: Australia                      2021     1746.0  
4       AUS: Australia                      2021     1863.0  
5       AUS: Australia                      2021      375.0  
6       AUS: Australia                      2021        0.8  
7       AUS: Australia                      2021        2.5  
8   1: New South Wales                      2021       46.0  
9   1: New South Wales                      2021       35.0  
10  1: New South Wales                      2021       38.0  
11  1: New South Wales                      2021       44.0  
12  1: New South Wales                      2021       38.0  
13  1: New South Wales                      2021       45.0  
14  1: New South Wales                      2021       41.0

In [11]:
# Before filtering:
abs_df.shape

(23632, 7)

First, let's get state-wide statistics for imputation purposes later. Since we are focusing on median personal income, we will only get each state's median personal income number.

In [12]:
stat = "2: Median total personal income ($/weekly)"

NSW_median_income = float(abs_df[(abs_df["MEDAVG: Median/Average"] == stat) &
                           (abs_df["REGION: Region"] == "1: New South Wales")]["OBS_VALUE"])

VIC_median_income = float(abs_df[(abs_df["MEDAVG: Median/Average"] == stat) &
                           (abs_df["REGION: Region"] == "2: Victoria")]["OBS_VALUE"])

QLD_median_income = float(abs_df[(abs_df["MEDAVG: Median/Average"] == stat) &
                           (abs_df["REGION: Region"] == "3: Queensland")]["OBS_VALUE"])

SA_median_income = float(abs_df[(abs_df["MEDAVG: Median/Average"] == stat) &
                           (abs_df["REGION: Region"] == "4: South Australia")]["OBS_VALUE"])

WA_median_income = float(abs_df[(abs_df["MEDAVG: Median/Average"] == stat) &
                           (abs_df["REGION: Region"] == "5: Western Australia")]["OBS_VALUE"])

TAS_median_income = float(abs_df[(abs_df["MEDAVG: Median/Average"] == stat) &
                           (abs_df["REGION: Region"] == "6: Tasmania")]["OBS_VALUE"])

NT_median_income = float(abs_df[(abs_df["MEDAVG: Median/Average"] == stat) &
                           (abs_df["REGION: Region"] == "7: Northern Territory")]["OBS_VALUE"])

ACT_median_income = float(abs_df[(abs_df["MEDAVG: Median/Average"] == stat) &
                           (abs_df["REGION: Region"] == "8: Australian Capital Territory")]["OBS_VALUE"])

Firstly, we only want SA2 data, and data from 2021.

In [13]:
abs_filtered = abs_df[(abs_df["REGION_TYPE: Region Type"] == "SA2: Statistical Area Level 2") & 
                      (abs_df["TIME_PERIOD: Time Period"] == 2021)]
abs_filtered.shape

(19722, 7)

In [14]:
abs_filtered.head(5)

DATAFLOW    MEDAVG: Median/Average  \
34  ABS:C21_G02_SA2(1.0.0)  1: Median age of persons   
35  ABS:C21_G02_SA2(1.0.0)  1: Median age of persons   
36  ABS:C21_G02_SA2(1.0.0)  1: Median age of persons   
37  ABS:C21_G02_SA2(1.0.0)  1: Median age of persons   
38  ABS:C21_G02_SA2(1.0.0)  1: Median age of persons   

                             REGION: Region       REGION_TYPE: Region Type  \
34                         101031014: Cooma  SA2: Statistical Area Level 2   
35            101051540: Goulburn Surrounds  SA2: Statistical Area Level 2   
36  102021052: Summerland Point - Gwandalan  SA2: Statistical Area Level 2   
37         103031075: Wollangambe - Wollemi  SA2: Statistical Area Level 2   
38                    105031106: Wellington  SA2: Statistical Area Level 2   

          STATE: State  TIME_PERIOD: Time Period  OBS_VALUE  
34  1: New South Wales                      2021       44.0  
35  1: New South Wales                      2021       48.0  
36  1: New South Wales                      2021       45.0  
37  1: New South Wales                      2021        0.0  
38  1: New South Wales                      2021       41.0

Now, we can convert the SA2 region into only its code, and also remove unnecessary features. We will also rename the relevant columns for ease of access.

In [15]:
# change SA2 feature
abs_filtered["REGION: Region"] = abs_filtered["REGION: Region"].str[:9].astype("int64")
abs_filtered.head(5)

/var/folders/cg/w3sqdjkn6dddbfbdcn4ckggh0000gn/T/ipykernel_70845/4185616500.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  abs_filtered["REGION: Region"] = abs_filtered["REGION: Region"].str[:9].astype("int64")


DATAFLOW    MEDAVG: Median/Average  REGION: Region  \
34  ABS:C21_G02_SA2(1.0.0)  1: Median age of persons       101031014   
35  ABS:C21_G02_SA2(1.0.0)  1: Median age of persons       101051540   
36  ABS:C21_G02_SA2(1.0.0)  1: Median age of persons       102021052   
37  ABS:C21_G02_SA2(1.0.0)  1: Median age of persons       103031075   
38  ABS:C21_G02_SA2(1.0.0)  1: Median age of persons       105031106   

         REGION_TYPE: Region Type        STATE: State  \
34  SA2: Statistical Area Level 2  1: New South Wales   
35  SA2: Statistical Area Level 2  1: New South Wales   
36  SA2: Statistical Area Level 2  1: New South Wales   
37  SA2: Statistical Area Level 2  1: New South Wales   
38  SA2: Statistical Area Level 2  1: New South Wales   

    TIME_PERIOD: Time Period  OBS_VALUE  
34                      2021       44.0  
35                      2021       48.0  
36                      2021       45.0  
37                      2021        0.0  
38                      2021       41.0

In [16]:
abs_filtered = abs_filtered.drop(["DATAFLOW", "REGION_TYPE: Region Type", "STATE: State",
                                  "TIME_PERIOD: Time Period"], axis=1) # drop useless columns
abs_filtered = abs_filtered.rename(columns={"MEDAVG: Median/Average": "statistic",
                                            "REGION: Region": "region",
                                            "OBS_VALUE": "value"}) # rename columns
abs_filtered.shape

(19722, 3)

We should engineer new columns based on the categorical values of the `Statistic` column.

In [17]:
abs_filtered = abs_filtered.pivot_table(index="region", columns="statistic", values="value", aggfunc="sum")
abs_filtered = abs_filtered.reset_index()
abs_filtered

statistic     region  1: Median age of persons  \
0          101021007                      51.0   
1          101021008                      38.0   
2          101021009                      37.0   
3          101021010                      36.0   
4          101021012                      37.0   
...              ...                       ...   
2467       901021002                      40.0   
2468       901031003                      38.0   
2469       901041004                      50.0   
2470       997979799                       0.0   
2471       999999499                       0.0   

statistic  2: Median total personal income ($/weekly)  \
0                                               760.0   
1                                               975.0   
2                                               996.0   
3                                              1104.0   
4                                              1357.0   
...                                               ...   
2467                                            741.0   
2468                                            585.0   
2469                                            736.0   
2470                                              0.0   
2471                                              0.0   

statistic  3: Median total family income ($/weekly)  \
0                                            1886.0   
1                                            2334.0   
2                                            2233.0   
3                                            2412.0   
4                                            3332.0   
...                                             ...   
2467                                         1678.0   
2468                                         1687.0   
2469                                         1630.0   
2470                                            0.0   
2471                                            NaN   

statistic  4: Median total household income ($/weekly)  \
0                                               1429.0   
1                                               1989.0   
2                                               1703.0   
3                                               1796.0   
4                                               3014.0   
...                                                ...   
2467                                            2519.0   
2468                                            1825.0   
2469                                            1184.0   
2470                                               0.0   
2471                                               NaN   

statistic  5: Median mortgage repayment ($/monthly)  \
0                                            1732.0   
1                                            1950.0   
2                                            1700.0   
3                                            1700.0   
4                                            2300.0   
...                                             ...   
2467                                         1972.0   
2468                                         1849.0   
2469                                         1300.0   
2470                                            0.0   
2471                                            NaN   

statistic  6: Median rent ($/weekly)  \
0                              330.0   
1                              350.0   
2                              330.0   
3                              310.0   
4                              430.0   
...                              ...   
2467                           231.0   
2468                            90.0   
2469                           240.0   
2470                             0.0   
2471                             NaN   

statistic  7: Average number of persons per bedroom  8: Average household size  
0                                               0.8                        2.2  
1                                               0.8 

In [18]:
# rename all columns for simplicity
abs_filtered.columns = ["region", "median_age", "median_personal_income",
                        "median_family_income", "median_household_income", "median_mortgage",
                        "median_rent", "avg_bedroom", "avg_household"]
abs_filtered.shape

(2472, 9)

In [19]:
abs_filtered.drop("region", axis=1).describe()

median_age  median_personal_income  median_family_income  \
count  2472.000000             2472.000000           2463.000000   
mean     39.728964              835.014159           2115.051969   
std       8.617423              286.225602            745.723936   
min       0.000000                0.000000              0.000000   
25%      35.000000              675.000000           1655.500000   
50%      39.000000              798.000000           2079.000000   
75%      44.000000              959.250000           2477.000000   
max      91.000000             3250.000000           7000.000000   

       median_household_income  median_mortgage  median_rent  avg_bedroom  \
count              2463.000000      2463.000000  2463.000000  2463.000000   
mean               1762.125863      1805.879009   358.440926     0.811246   
std                 621.424829       688.744319   129.056917     0.172310   
min                   0.000000         0.000000     0.000000     0.000000   
25%                1346.000000      1430.000000   290.000000     0.800000   
50%                1735.000000      1815.000000   365.000000     0.800000   
75%                2135.500000      2167.000000   426.000000     0.900000   
max                7000.000000      9999.000000  1200.000000     2.000000   

       avg_household  
count    2463.000000  
mean        2.508242  
std         0.537332  
min         0.000000  
25%         2.300000  
50%         2.500000  
75%         2.800000  
max         6.000000

We should remove instances which include outlier values for some features. For example, we can see that minimum values for a lot of the statistics are zero, which doesn't make sense. We should also remove any NaN values.

In [20]:
# remove outliers
abs_filtered = abs_filtered[(abs_filtered["median_age"] > 0) & 
                            (abs_filtered["median_personal_income"] > 0) &
                            (abs_filtered["median_family_income"] > 0) &
                            (abs_filtered["median_household_income"] > 0) &
                            (abs_filtered["median_mortgage"] > 0) &
                            (abs_filtered["median_rent"] > 0) &
                            (abs_filtered["avg_bedroom"] > 0) &
                            (abs_filtered["avg_household"] > 0)]

abs_filtered = abs_filtered.dropna() # remove NaN values

abs_filtered.drop("region", axis=1).describe()

median_age  median_personal_income  median_family_income  \
count  2350.000000             2350.000000           2350.000000   
mean     40.180851              845.071064           2167.366809   
std       6.407814              235.630911            639.611874   
min      23.000000              240.000000            551.000000   
25%      36.000000              684.250000           1695.500000   
50%      39.000000              803.000000           2094.500000   
75%      44.000000              957.750000           2487.000000   
max      70.000000             2150.000000           7000.000000   

       median_household_income  median_mortgage  median_rent  avg_bedroom  \
count              2350.000000      2350.000000  2350.000000  2350.000000   
mean               1799.736170      1881.170213   370.586383     0.822340   
std                 540.384664       585.246160   112.146069     0.117765   
min                 575.000000       109.000000    20.000000     0.500000   
25%                1373.000000      1500.000000   300.000000     0.800000   
50%                1755.000000      1849.000000   370.000000     0.800000   
75%                2148.750000      2167.000000   430.000000     0.900000   
max                4799.000000      9999.000000  1200.000000     1.900000   

       avg_household  
count    2350.000000  
mean        2.560596  
std         0.371431  
min         1.300000  
25%         2.300000  
50%         2.500000  
75%         2.800000  
max         5.100000

We can see that the distributions of these numeric features now look much more sensible.

In [21]:
# external dataset shape after filtering
abs_filtered.shape

(2350, 9)

Now we can take the median personal income (our variable of interest) and merge this with our customer dataset according to SA2 region code.

In [23]:
median_personal_income_df = abs_filtered.loc[:, ["region", "median_personal_income"]]
customers_merged["SA2_CODE_2021"] = customers_merged["SA2_CODE_2021"].astype("int") # helps with merging

df_merged = pd.merge(customers_merged, median_personal_income_df, left_on="SA2_CODE_2021", right_on="region", how="left")
df_merged = df_merged.drop("SA2_CODE_2021", axis=1) # drop duplicate column
df_merged.head(5)

name                        address state  postcode  gender  \
0   Yolanda Williams     413 Haney Gardens Apt. 742    WA      6935  Female   
1         Mary Smith                3764 Amber Oval   NSW      2782  Female   
2      Jill Jones MD             40693 Henry Greens    NT       862  Female   
3    Lindsay Jimenez      00653 Davenport Crossroad   NSW      2780  Female   
4  Rebecca Blanchard  9271 Michael Manors Suite 651    WA      6355  Female   

   consumer_id       region  median_personal_income  
0      1195503          NaN                     NaN  
1       179208  124011455.0                   740.0  
2      1194530  702021055.0                   416.0  
3       154128  124011452.0                   687.0  
4       712975  509031247.0                   897.0

Since we are missing some income numbers (see the `NaN` value), we will impute each missing value with the median income of the state where the customer is from.

In [24]:
replacement_values = {"NSW": NSW_median_income, # values determined through initial ABS dataset
                      "VIC": VIC_median_income,
                      "QLD": QLD_median_income,
                      "SA": SA_median_income,
                      "WA": WA_median_income,
                      "TAS": TAS_median_income,
                      "NT": NT_median_income,
                      "ACT": ACT_median_income}

replacement_series = df_merged["state"].map(replacement_values)

df_merged["median_personal_income"] = df_merged["median_personal_income"].fillna(replacement_series)
df_merged.head(5)

name                        address state  postcode  gender  \
0   Yolanda Williams     413 Haney Gardens Apt. 742    WA      6935  Female   
1         Mary Smith                3764 Amber Oval   NSW      2782  Female   
2      Jill Jones MD             40693 Henry Greens    NT       862  Female   
3    Lindsay Jimenez      00653 Davenport Crossroad   NSW      2780  Female   
4  Rebecca Blanchard  9271 Michael Manors Suite 651    WA      6355  Female   

   consumer_id       region  median_personal_income  
0      1195503          NaN                   848.0  
1       179208  124011455.0                   740.0  
2      1194530  702021055.0                   416.0  
3       154128  124011452.0                   687.0  
4       712975  509031247.0                   897.0

In [25]:
df_merged["median_personal_income"].isna().any()

False

We can see that our median personal income column now has no missing values, and hence the data has been properly imputed.

In [28]:
# define the folder path and filename
output_path = "../data/curated/"
file_name = "consumers_median_income"

# create the folder if it doesn't exist
if not os.path.exists(output_path):
    os.makedirs(output_path)

# save df to csv in the specified folder
file_path = os.path.join(output_path, file_name)
df_merged.to_csv(file_path, index=False)